'''
1
'''

In [1]:
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler, PowerTransformer, QuantileTransformer, MinMaxScaler, MaxAbsScaler, RobustScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

In [2]:
dataset = pd.read_csv('auto.csv')

In [6]:
dataset.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...   engine-size  \
0  convertible          rwd           front        88.6  ...           130   
1  convertible          rwd           front        88.6  ...           130   
2    hatchback          rwd           front        94.5  ...           152   
3        sedan          fwd           front        99.8  ...           109   
4        sedan          4wd           front        99.4  ...           136   
5        sedan          fwd           front        99.8  ...           136   
6        sedan          fwd           front       105.8  ...           136   
7        wagon          fwd           front       105.8  ...           136   
8        sedan          fwd           front       105.8  ...           131   
9    hatchback          4wd           front        99.5  ...           131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

In [5]:
dataset.describe()

symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight   engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000    205.000000         205.000000  205.000000   205.000000  
mean   2555.565854    126.907317          10.142537   25.219512    30.751220  
std     520.680204     41.642693           3.972040    6.542142     6.886443  
min    1488.000000     61.000000           7.000000   13.000000    16.000000  
25%    2145.000000     97.000000           8.600000   19.000000    25.000000  
50%    2414.000000    120.000000           9.000000   24.000000    30.000000  
75%    2935.000000    141.000000           9.400000   30.000000    34.000000  
max    4066.000000    326.000000          23.000000   49.000000    54.000000

In [10]:
dataset.replace('?', np.NaN, inplace=True)

In [11]:
dataset.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2               NaN         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1               NaN         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0               NaN         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...   engine-size  \
0  convertible          rwd           front        88.6  ...           130   
1  convertible          rwd           front        88.6  ...           130   
2    hatchback          rwd           front        94.5  ...           152   
3        sedan          fwd           front        99.8  ...           109   
4        sedan          4wd           front        99.4  ...           136   
5        sedan          fwd           front        99.8  ...           136   
6        sedan          fwd           front       105.8  ...           136   
7        wagon          fwd           front       105.8  ...           136   
8        sedan          fwd           front       105.8  ...           131   
9    hatchback          4wd           front        99.5  ...           131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22    NaN  

[10 rows x 26 columns]

In [17]:
dataset.isnull().sum().sort_values(ascending=False)

normalized-losses    41
price                 4
stroke                4
bore                  4
peak-rpm              2
horsepower            2
num-of-doors          2
length                0
make                  0
fuel-type             0
aspiration            0
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
height                0
width                 0
highway-mpg           0
curb-weight           0
engine-type           0
num-of-cylinders      0
 engine-size          0
fuel-system           0
compression-ratio     0
city-mpg              0
symboling             0
dtype: int64

In [4]:
dataset.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
 engine-size           int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

In [25]:
#changing data types
dataset['normalized-losses'] = dataset['normalized-losses'].astype(float)
dataset['price'] = dataset['price'].astype(float)
dataset['bore'] = dataset['bore'].astype(float)
dataset['stroke'] = dataset['stroke'].astype(float)
dataset['peak-rpm'] = dataset['peak-rpm'].astype(float)
dataset['horsepower'] = dataset['horsepower'].astype(float)

In [35]:
'''get their mean'''
avg_norm_loss = dataset['normalized-losses'].mean(axis=0)
avg_prc = dataset['price'].mean(axis=0)
avg_bo = dataset['bore'].mean(axis=0)
avg_st = dataset['stroke'].mean(axis=0)
avg_rpm = dataset['peak-rpm'].mean(axis=0)
avg_hp = dataset['horsepower'].mean(axis=0)

dataset['normalized-losses'].replace(np.nan, avg_norm_loss, inplace=True)
dataset['price'].replace(np.nan, avg_prc, inplace=True)
dataset['bore'].replace(np.nan, avg_bo, inplace=True)
dataset['stroke'].replace(np.nan, avg_st, inplace=True)
dataset['peak-rpm'].replace(np.nan, avg_rpm, inplace=True)
dataset['horsepower'].replace(np.nan, avg_hp, inplace=True)

In [37]:
dataset['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

In [38]:
dataset['num-of-doors'].replace(np.nan, 'four', inplace=True)

In [47]:
#ganti huruf jadi angka
cleanup_nums = {
    'num-of-doors' :{
        'two' : 2,
        'four' : 4
    },
    'num-of-cylinders':{
        'two' : 2,
        'three' : 3,
        'four' : 4,
        'five' : 5,
        'six' : 6,
        'eight' : 8,
        'twelve' : 12
    }
}

dataset.replace(cleanup_nums, inplace=True)

In [39]:
dataset.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
 engine-size         0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

In [42]:
#Feature Engineering
dataset['make'].value_counts()

toyota           32
nissan           18
mazda            17
honda            13
mitsubishi       13
volkswagen       12
subaru           12
volvo            11
peugot           11
dodge             9
mercedes-benz     8
bmw               8
plymouth          7
audi              7
saab              6
porsche           5
isuzu             4
chevrolet         3
jaguar            3
alfa-romero       3
renault           2
mercury           1
Name: make, dtype: int64

In [45]:
make_categorized = pd.get_dummies(dataset['make'])
dataset = pd.concat([dataset, make_categorized], axis=1)

In [49]:
dataset['fuel-type'].value_counts()

gas       185
diesel     20
Name: fuel-type, dtype: int64

In [50]:
fuel_categorized = pd.get_dummies(dataset['fuel-type'])
dataset = pd.concat([dataset, fuel_categorized], axis=1)

In [53]:
dataset.drop(['make', 'fuel-type'], axis=1, inplace=True)

In [55]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 48 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    float64
 2   aspiration         205 non-null    object 
 3   num-of-doors       205 non-null    int64  
 4   body-style         205 non-null    object 
 5   drive-wheels       205 non-null    object 
 6   engine-location    205 non-null    object 
 7   wheel-base         205 non-null    float64
 8   length             205 non-null    float64
 9   width              205 non-null    float64
 10  height             205 non-null    float64
 11  curb-weight        205 non-null    int64  
 12  engine-type        205 non-null    object 
 13  num-of-cylinders   205 non-null    int64  
 14   engine-size       205 non-null    int64  
 15  fuel-system        205 non-null    object 
 16  bore               205 non

In [56]:
features = dataset.drop(['aspiration','body-style','drive-wheels','engine-location','engine-type','fuel-system', 'price'], axis=1)
label = dataset['price']

In [57]:
import sys
scalers = [StandardScaler(), PowerTransformer(method='yeo-johnson'), PowerTransformer(method='box-cox'), QuantileTransformer(), MinMaxScaler(), MaxAbsScaler(), RobustScaler()]
for i in scalers:
    try:
        scaler = i
        features_scaled = scaler.fit_transform(features.values)
        label_scaled = scaler.fit_transform(label.values.reshape(-1,1))
        feature_train, feature_test,label_train, label_test = train_test_split(features_scaled, label_scaled, random_state=10)
                    
        lm = LinearRegression()
        lm.fit(feature_train, label_train)
        predictions = lm.predict(feature_test)

        mse = mean_squared_error(label_test, predictions)
        mae = mean_absolute_error(label_test, predictions)
        r2 = r2_score(label_test, predictions)
        print("Preprocessing by {}".format(i))
        print("MSE (Mean Squared Error)", round(mse, 3))
        print("MAE (Mean Absolute Error)", round(mae, 3))
        print("r^2 score", round(r2,3))
        print('RMSE (Root Mean Squared Error):', round(np.sqrt(mean_squared_error(label_test, predictions)), 3))
        print("")
    except:
        print("Unexpected error:", sys.exc_info()[0])
        print("")
        pass

Preprocessing by StandardScaler()
MSE (Mean Squared Error) 0.321
MAE (Mean Absolute Error) 0.315
r^2 score 0.622
RMSE (Root Mean Squared Error): 0.566

Preprocessing by PowerTransformer()
MSE (Mean Squared Error) 0.0
MAE (Mean Absolute Error) 0.0
r^2 score 1.0
RMSE (Root Mean Squared Error): 0.0

Unexpected error: <class 'ValueError'>

Preprocessing by QuantileTransformer()
MSE (Mean Squared Error) 0.013
MAE (Mean Absolute Error) 0.079
r^2 score 0.805
RMSE (Root Mean Squared Error): 0.115

Preprocessing by MinMaxScaler()
MSE (Mean Squared Error) 0.012
MAE (Mean Absolute Error) 0.062
r^2 score 0.621
RMSE (Root Mean Squared Error): 0.111

Preprocessing by MaxAbsScaler()
MSE (Mean Squared Error) 0.01
MAE (Mean Absolute Error) 0.055
r^2 score 0.621
RMSE (Root Mean Squared Error): 0.098

Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.261
MAE (Mean Absolute Error) 0.286
r^2 score 0.621
RMSE (Root Mean Squared Error): 0.511



Metode yeo-johnson memiliki nilai r^2 score sempurna (1) dan rmse 0. Tetapi metode ini tidak optimal karena it tends to be overfitted. Metode preprocessing by QuantileTransformer() memiliki nilai r^2 score yang paling tinggi (0.805) dan nilai rmse yang cukup rendah (0.115) dan metode ini lebih optimal karena tidak overfitting. Metode box-cox error karena metode ini hanya dapat bekerja di dataset yang positif.

'''
2
'''

In [73]:
#Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor
est_num = [*range(10, 201, 10)]
for index, val in enumerate(est_num):
    rfg = RandomForestRegressor(n_estimators = val, bootstrap = True, max_features = 'sqrt')
    rfg.fit(feature_train, label_train)
    y_pred = rfg.predict(feature_test)
    print('Accuracy of Random Forest with estimator {} on test set: {:.2f}'.format(val, rfg.score(feature_test, label_test)))
    mse = mean_squared_error(label_test, y_pred)
    mae = mean_absolute_error(label_test, y_pred)
    r2 = r2_score(label_test, y_pred)
    print("Preprocessing by {}".format(i))
    print("MSE (Mean Squared Error)", round(mse, 3))
    print("MAE (Mean Absolute Error)", round(mae, 3))
    print("r^2 score", round(r2,3))
    print('RMSE (Root Mean Squared Error):', round(np.sqrt(mean_squared_error(label_test, y_pred)), 3))
    print("")

Accuracy of Random Forest with estimator 10 on test set: 0.90
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.094
MAE (Mean Absolute Error) 0.204
r^2 score 0.904
RMSE (Root Mean Squared Error): 0.307

Accuracy of Random Forest with estimator 20 on test set: 0.89
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.107
MAE (Mean Absolute Error) 0.225
r^2 score 0.891
RMSE (Root Mean Squared Error): 0.327

Accuracy of Random Forest with estimator 30 on test set: 0.87
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.125
MAE (Mean Absolute Error) 0.248
r^2 score 0.872
RMSE (Root Mean Squared Error): 0.354

Accuracy of Random Forest with estimator 40 on test set: 0.88
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.114
MAE (Mean Absolute Error) 0.227
r^2 score 0.884
RMSE (Root Mean Squared Error): 0.338

Accuracy of Random Forest with estimator 50 on test set: 0.90
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.102
MAE (Mean Absolute Error) 0

In [72]:
#XGBoost
import xgboost as xgb
est_num = [*range(10, 201, 10)]
for index, val in enumerate(est_num):
    xgreg = xgb.XGBRegressor(objective='reg:squarederror', colsample_bytree=1, eta=0.3, learning_rate=0.1, max_depth=5, alpha=3, n_estimators = val)
    xgreg.fit(feature_train, label_train)
    y_pred = xgreg.predict(feature_test)
    print('Accuracy of xgboost with estimator {} on test set: {:.2f}'.format(val, xgreg.score(feature_test, label_test)))
    mse = mean_squared_error(label_test, y_pred)
    mae = mean_absolute_error(label_test, y_pred)
    r2 = r2_score(label_test, y_pred)
    print("Preprocessing by {}".format(i))
    print("MSE (Mean Squared Error)", round(mse, 3))
    print("MAE (Mean Absolute Error)", round(mae, 3))
    print("r^2 score", round(r2,3))
    print('RMSE (Root Mean Squared Error):', round(np.sqrt(mean_squared_error(label_test, y_pred)), 3))
    print("")

Accuracy of xgboost with estimator 10 on test set: 0.72
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.273
MAE (Mean Absolute Error) 0.433
r^2 score 0.722
RMSE (Root Mean Squared Error): 0.523

Accuracy of xgboost with estimator 20 on test set: 0.86
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.135
MAE (Mean Absolute Error) 0.294
r^2 score 0.863
RMSE (Root Mean Squared Error): 0.367

Accuracy of xgboost with estimator 30 on test set: 0.88
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.114
MAE (Mean Absolute Error) 0.251
r^2 score 0.884
RMSE (Root Mean Squared Error): 0.338

Accuracy of xgboost with estimator 40 on test set: 0.89
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.112
MAE (Mean Absolute Error) 0.239
r^2 score 0.886
RMSE (Root Mean Squared Error): 0.335

Accuracy of xgboost with estimator 50 on test set: 0.89
Preprocessing by RobustScaler()
MSE (Mean Squared Error) 0.106
MAE (Mean Absolute Error) 0.231
r^2 score 0.892
RMSE (Roo

In [70]:
#SVR
from sklearn.svm import SVR

scaler = StandardScaler()
features_scaled = scaler.fit_transform(features.values)
label_scaled = scaler.fit_transform(label.values.reshape(-1,1))
feature_train, feature_test,label_train, label_test = train_test_split(features_scaled, label_scaled)

kernels = ['rbf', 'poly', 'linear']
for kernel in kernels:
    svreg = SVR(kernel=kernel)
    svreg.fit(feature_train, label_train)
    y_pred = svreg.predict(feature_test)
    print('Accuracy of SVR with kernel {} on test set: {:.2f}'.format(kernel, svreg.score(feature_test, label_test)))
    mse = mean_squared_error(label_test, y_pred)
    mae = mean_absolute_error(label_test, y_pred)
    r2 = r2_score(label_test, y_pred)
    print("Preprocessing by {}".format(kernel))
    print("MSE (Mean Squared Error)", round(mse, 3))
    print("MAE (Mean Absolute Error)", round(mae, 3))
    print("r^2 score", round(r2,3))
    print('RMSE (Root Mean Squared Error):', round(np.sqrt(mean_squared_error(label_test, y_pred)), 3))
    print("")

Accuracy of SVR with kernel rbf on test set: 0.86
Preprocessing by rbf
MSE (Mean Squared Error) 0.141
MAE (Mean Absolute Error) 0.236
r^2 score 0.857
RMSE (Root Mean Squared Error): 0.375

Accuracy of SVR with kernel poly on test set: 0.88
Preprocessing by poly
MSE (Mean Squared Error) 0.12
MAE (Mean Absolute Error) 0.266
r^2 score 0.878
RMSE (Root Mean Squared Error): 0.346

Accuracy of SVR with kernel linear on test set: 0.84
Preprocessing by linear
MSE (Mean Squared Error) 0.154
MAE (Mean Absolute Error) 0.254
r^2 score 0.844
RMSE (Root Mean Squared Error): 0.392



In [69]:
#kNNRegressors
from sklearn.neighbors import KNeighborsRegressor
n_neighbors = [*range(1, 20, 1)]
for index,val in enumerate(n_neighbors):
    kreg = KNeighborsRegressor(n_neighbors=val, weights='distance', algorithm='ball_tree')
    kreg.fit(feature_train, label_train)
    y_pred = kreg.predict(feature_test)
    print('Accuracy of kNNRegressors with n_neighbor {} on test set: {:.2f}'.format(val, kreg.score(feature_test, label_test)))
    mse = mean_squared_error(label_test, y_pred)
    mae = mean_absolute_error(label_test, y_pred)
    r2 = r2_score(label_test, y_pred)
    print("Preprocessing by {}".format(kernel))
    print("MSE (Mean Squared Error)", round(mse, 3))
    print("MAE (Mean Absolute Error)", round(mae, 3))
    print("r^2 score", round(r2,3))
    print('RMSE (Root Mean Squared Error):', round(np.sqrt(mean_squared_error(label_test, y_pred)), 3))
    print("")

Accuracy of kNNRegressors with n_neighbor 1 on test set: 0.63
Preprocessing by linear
MSE (Mean Squared Error) 0.314
MAE (Mean Absolute Error) 0.307
r^2 score 0.63
RMSE (Root Mean Squared Error): 0.56

Accuracy of kNNRegressors with n_neighbor 2 on test set: 0.74
Preprocessing by linear
MSE (Mean Squared Error) 0.222
MAE (Mean Absolute Error) 0.261
r^2 score 0.739
RMSE (Root Mean Squared Error): 0.471

Accuracy of kNNRegressors with n_neighbor 3 on test set: 0.80
Preprocessing by linear
MSE (Mean Squared Error) 0.173
MAE (Mean Absolute Error) 0.247
r^2 score 0.797
RMSE (Root Mean Squared Error): 0.416

Accuracy of kNNRegressors with n_neighbor 4 on test set: 0.83
Preprocessing by linear
MSE (Mean Squared Error) 0.147
MAE (Mean Absolute Error) 0.238
r^2 score 0.826
RMSE (Root Mean Squared Error): 0.384

Accuracy of kNNRegressors with n_neighbor 5 on test set: 0.83
Preprocessing by linear
MSE (Mean Squared Error) 0.142
MAE (Mean Absolute Error) 0.245
r^2 score 0.833
RMSE (Root Mean Squar

Model yang paling ideal untuk dataset ini adalah model RandomForestRegressor dengan nilai r^2 score tertinggi dan rmse terendah dibandingkan model xgboost, svr, serta kNNregressor. Nilai r^2 score yang diperoleh dengan n_estimator 180 adalah 0.91 dan rmse sebesar 0.303. Jika dibandingkan dengan data-data pada boston-housing dan winequality-white model RandomForestRegressor memiliki tingkat akurasi yang tinggi dengan nilai standar-deviasi yang rendah.